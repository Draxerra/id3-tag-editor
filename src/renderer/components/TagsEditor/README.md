# Tags Editor

## Functionality

The tags editor is a form component, which takes an array of tags and groups them together into one for editing. Each tag element passed in will automatically be displayed as a text field, but may be overwritten using a named slot.
`<keep>` is displayed when key values are different in the array provided. `<keep>` will get interpolated by the component on save, and will return the original value.
Regex's may also be entered into the form, such as: `</Original Mix/g>`. Regex's will get interpolated by the component on save, and will return the result of the regex, if any. If no result is found, the original value will instead be returned.

### Props

`<tags-editor>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `equality-check-fn` | (prev, curr) => boolean *optional* | A function to determine whether two tag values are equal. Defaults to `(prev, curr) => prev === curr`.
| `items` | object[] | The array of tags.
| `headers` | IHeader[] | The array of headers to display in the form. Should be in the form `{ text: string, value: string }` where text is the label to display, and value is the key it corresponds to in the items array.

### Events

`<tags-editor>` emits the following events:

| event | args type | description
| --- | --- | ---
| `submit` | object[] | The updated array of tags. Emitted upon form submission.

### Slots

`<tags-editor>` exposes the followings slots:

| slot | props | description
| --- | --- | ---
| `{header value}` | form: object[], header: IHeader | Optional slot for custom inputs. The name of the slot will be a header value provided in the headers prop. Form contains the key/value pairs of all the headers. Header contains the current tag header being iterated over.

### Example

```html
<tags-editor :items="items"
             :headers="headers"
             :equality-check-fn="equalityCheckFn"
             @submit="submit">
    <div slot="artist" slot-scope="{ form, header }">
        <label>{{ header.label }}</label>
        <input type="text"
               v-model="form[header.value]" />
    </div>
</tags-editor>
```

```ts
import Vue from 'vue';
import { Component } from "vue-property-decorator";

import { writeTagsSync } from "taglib2";

import TagsEditor from '~/src/renderer/components/TagsEditor/TagsEditor.vue';

@Component({
    components: {
        TagsEditor,
    },
})
export default class GenericComponent extends Vue {
    private items = [{
        title: "Test Song",
        artist: "Test Artist",
        path: "/path/to/file.mp3",
    }, {
        title: "Test Song 2",
        artist: "Test Artist",
        path: "/path/to/file.mp3",
    }];

    private headers = [{
        text: "Title",
        value: "title",
    }, {
        text: "Artist",
        value: "artist",
    }];

    private equalityCheckFn(prev, curr) {
        return prev === curr;
    }

    private submit(updatedTags) {
        updatedTags.map((updatedTag) => writeTagsSync(updatedTag.path, updatedTag));
    }
}
```