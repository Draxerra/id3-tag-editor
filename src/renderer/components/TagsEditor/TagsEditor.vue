<template lang="pug">
    v-form.TagsEditor
        slot(v-for="header in headers" :name="header.value" :form="formTag" :header="header")
            v-text-field(:label="header.text"
                         v-model="formTag[header.value]"
                         :id="header.value").TagsEditorTextField
        v-btn(@click="submit" color="primary").TagsEditorSubmitBtn Save Changes
</template>

<script lang="ts">
import Vue from "vue";
import { Component, Prop, Watch } from "vue-property-decorator";

import { IHeader } from "~/src/types/header";

import VBtn from "vuetify/es5/components/VBtn";
import VForm from "vuetify/es5/components/VForm";
import VTextField from "vuetify/es5/components/VTextField";

@Component({
    components: {
        VBtn,
        VForm,
        VTextField,
    },
})
export default class TagsEditor extends Vue {
    @Prop({default: (prev, curr) => prev.toString() === curr.toString(),
           type: Function}) private equalityCheckFn!: (prev, curr) => boolean;
    @Prop({type: Array, required: true}) private items!: Array<{[key: string]: any}>;
    @Prop({type: Array, required: true}) private headers!: IHeader[];
    private regexFinder = /<\/([^>]*)\/([a-zA-Z1-9]*)>/g;

    private get groupedItems() {
        return this.headers.reduce((acc, tag) => {
            acc[tag.value] = this.items.map((item) => item[tag.value]);
            return acc;
        }, {} as {[key: string]: any[]});
    }

    private formTag: {[key: string]: any} = {};

    @Watch("groupedItems", { immediate: true })
    private groupedItemsChanged() {
        this.formTag = Object.entries(this.groupedItems).reduce((acc, [key, val]) => {
            acc[key] = val.reduce((valAcc, currVal) => {
                return this.equalityCheckFn(valAcc, currVal) ? valAcc : "<keep>";
            });
            return acc;
        }, {} as {[key: string]: any});
    }

    private replaceKeep(val: string | number | object, orig: string | number | object) {
        const replaced = typeof orig === "object" ? orig : val.toString().replace(/<keep>/g, orig.toString());
        const parsed = parseInt(replaced.toString(), 10);
        return typeof orig === "number" && typeof replaced === "string" && !isNaN(parsed) ? parsed : replaced;
    }

    private replaceRegex(val: string | number | object, orig: string | number | object) {
        const regexs = val.toString().match(new RegExp(this.regexFinder));
        const replaced = !regexs ? val : regexs.reduce((acc, regex) => {
            const [, pattern, flags] = new RegExp(this.regexFinder).exec(regex) as RegExpExecArray;
            const matches = orig.toString().match(new RegExp(pattern, flags));
            return acc.toString().replace(regex, matches ? matches.join("") : "");
        }, val) || orig;
        const parsed = parseInt(replaced.toString(), 10);
        return typeof orig === "number" && typeof replaced === "string" && !isNaN(parsed) ? parsed : replaced;
    }

    private submit() {
        const updatedTags = this.items.map((currTag) => {
            return Object.entries(currTag).reduce((acc, [key, val]) => {
                const formVal = this.formTag[key] === undefined ? String(acc[key]) : String(this.formTag[key]);
                acc[key] = typeof this.formTag[key] === "object" ?
                    this.formTag[key] :
                    this.replaceRegex(this.replaceKeep(formVal, val), val);
                return acc;
            }, {...currTag});
        });
        this.$emit("submit", updatedTags);
    }
}
</script>
