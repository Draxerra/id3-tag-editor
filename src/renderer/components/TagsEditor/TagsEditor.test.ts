import { config, mount } from "@vue/test-utils";
import TagsEditor from "./TagsEditor.vue";

import Vue from "vue";
import Vuetify from "vuetify/es5/components/Vuetify";

Vue.use(Vuetify);
config.logModifiedComponents = false;

const item1 = {
    albumArtist: "Test Artist",
    artist: "Test Artist",
    name: "Test Song",
};
const item2 = {
    albumArtist: "Test Artist",
    artist: "Test Artist feat. Test Artist 2",
    name: "Test Song 2",
};

const items = [item1, item2];
const itemsWithArray = [{...item1, albumArtist: []}, {...item2, albumArtist: []}];

const header1 = {
    text: "Track Name",
    value: "name",
};
const header2 = {
    text: "Artist",
    value: "artist",
};
const header3 = {
    text: "Album Artist",
    value: "albumArtist",
};
const headers = [header1, header2, header3];

describe("Tags Editor Component", () => {

    it("is a Vue instance", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it("snapshot has not changed", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("displays a text input field for each header", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        expect(wrapper.find(".TagsEditorTextField #name").is("input")).toBeTruthy();
        expect(wrapper.find(".TagsEditorTextField #artist").is("input")).toBeTruthy();
        expect(wrapper.find(".TagsEditorTextField #albumArtist").is("input")).toBeTruthy();
    });

    it("displays a label for each header", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        const labels = wrapper.findAll(".TagsEditorTextField label");
        expect(labels.at(0).text()).toEqual("Track Name");
        expect(labels.at(1).text()).toEqual("Artist");
        expect(labels.at(2).text()).toEqual("Album Artist");
    });

    it("overrides text field input with named slot", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, scopedSlots: {
            artist: `<div slot-scope='props' class='TagsEditorCustomField'>
                        <label :for='props.header.value'>{{ props.header.text }}</label>
                        <input type='text'
                               id='artist'
                               :value='props.form[props.header.value]'
                               @input="props.form[props.header.value] = $event.target.value" />
                     </div>`,
        }, sync: false });
        const labels = wrapper.findAll(".TagsEditorTextField label");
        expect(labels.at(0).text()).toEqual("Track Name");
        expect(wrapper.find(".TagsEditorCustomField label").text()).toEqual("Artist");
        expect(labels.at(1).text()).toEqual("Album Artist");
    });

    it("displays keep with non unique values in items array", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        expect(wrapper.vm.$data.formTag.name).toEqual("<keep>");
        expect(wrapper.vm.$data.formTag.artist).toEqual("<keep>");
    });

    it("displays value with unique values in items array", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual("Test Artist");
    });

    it("updates form tag with updated value", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        expect(wrapper.vm.$data.formTag.name).toEqual("<keep>");
        const input = wrapper.find("#name");
        (input.element as HTMLInputElement).value = "Updated Name";
        input.trigger("input");
        expect(wrapper.vm.$data.formTag.name).toEqual("Updated Name");
    });

    it("updates form tag with updated slot value", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, scopedSlots: {
            artist: `<div slot-scope='props' class='TagsEditorCustomField'>
                        <label :for='props.header.value'>{{ props.header.text }}</label>
                        <input type='text'
                               id='artist'
                               :value='props.form[props.header.value]'
                               @input="props.form[props.header.value] = $event.target.value" />
                     </div>`,
        }, sync: false });
        expect(wrapper.vm.$data.formTag.artist).toEqual("<keep>");
        const input = wrapper.find("#artist");
        (input.element as HTMLInputElement).value = "Updated Artist";
        input.trigger("input");
        expect(wrapper.vm.$data.formTag.artist).toEqual("Updated Artist");
    });

    it("clicking submit button emits submit event", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.emitted("submit")).toEqual([[items]]);
    });

    it("clicking submit with <keep> value is transformed to original value", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });
        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.vm.$data.formTag.artist).toEqual("<keep>");
        expect(wrapper.emitted("submit")).toEqual([[items]]);
    });

    it("clicking submit without <keep> value is new value for each element in array", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });

        const input = wrapper.find("#albumArtist");
        (input.element as HTMLInputElement).value = "New Album Artist";
        input.trigger("input");

        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual("New Album Artist");
        expect(wrapper.emitted("submit")).toEqual([[[{
            ...item1,
            albumArtist: "New Album Artist",
        }, {
            ...item2,
            albumArtist: "New Album Artist",
        }]]]);
    });

    it("clicking submit with <keep> value on arrays and objects returns them", () => {
        const itemsWithDiffArray = [{...item1, albumArtist: [1]}, {...item2, albumArtist: [2]}];
        const wrapper = mount(TagsEditor, { propsData: { headers, items: itemsWithDiffArray }, sync: false });
        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual("<keep>");
        expect(wrapper.emitted("submit")).toEqual([[[{
            ...item1,
            albumArtist: [1],
        }, {
            ...item2,
            albumArtist: [2],
        }]]]);
    });

    it("clicking submit with <keep> value on numbers returns as a number", () => {
        const itemsWithDiffArray = [{...item1, albumArtist: 1}, {...item2, albumArtist: 2}];
        const wrapper = mount(TagsEditor, { propsData: { headers, items: itemsWithDiffArray }, sync: false });
        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual("<keep>");
        expect(wrapper.emitted("submit")).toEqual([[[{
            ...item1,
            albumArtist: 1,
        }, {
            ...item2,
            albumArtist: 2,
        }]]]);
    });

    it("clicking submit with regex value is evaluated as new value for each element in array", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });

        const input = wrapper.find("#albumArtist");
        (input.element as HTMLInputElement).value = "</Test/g>";
        input.trigger("input");

        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual("</Test/g>");
        expect(wrapper.emitted("submit")).toEqual([[[{
            ...item1,
            albumArtist: "Test",
        }, {
            ...item2,
            albumArtist: "Test",
        }]]]);
    });

    it(`clicking submit with regex is evaluated and all matches are joined together to
        form new value for each element in array`, () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });

        const input = wrapper.find("#albumArtist");
        (input.element as HTMLInputElement).value = "</t/ig>";
        input.trigger("input");

        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual("</t/ig>");
        expect(wrapper.emitted("submit")).toEqual([[[{
            ...item1,
            albumArtist: "Tttt",
        }, {
            ...item2,
            albumArtist: "Tttt",
        }]]]);
    });

    it("clicking submit with regex is evaluated and no matches uses default value", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items }, sync: false });

        const input = wrapper.find("#albumArtist");
        (input.element as HTMLInputElement).value = "</No Match/g>";
        input.trigger("input");

        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual("</No Match/g>");
        expect(wrapper.emitted("submit")).toEqual([[items]]);
    });

    it("clicking submit with regex returns a number", () => {
        const itemsWithNumber = [{...item1, albumArtist: 1000}, {...item2, albumArtist: 1000}];
        const wrapper = mount(TagsEditor, { propsData: { headers, items: itemsWithNumber }, sync: false });

        const input = wrapper.find("#albumArtist");
        (input.element as HTMLInputElement).value = "</1/g>";
        input.trigger("input");

        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual("</1/g>");
        expect(wrapper.emitted("submit")).toEqual([[[
            {...item1, albumArtist: 1},
            {...item2, albumArtist: 1},
        ]]]);
    });

    it("custom equality check fn prop correctly displays equal arrays and objects", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items: itemsWithArray }, sync: false });
        expect(wrapper.vm.$data.formTag.albumArtist).toEqual([]);
    });

    it("ignores arrays and objects on submit", () => {
        const wrapper = mount(TagsEditor, { propsData: { headers, items: itemsWithArray }, sync: false });
        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.emitted("submit")).toEqual([[itemsWithArray]]);
    });

    it("ignores keys not in header prop on submit", () => {
        const extraItems = [{...item1, compilation: 0}, {...item2, compilation: 1}];
        const wrapper = mount(TagsEditor, { propsData: { headers, items: extraItems }, sync: false });
        wrapper.find(".TagsEditorSubmitBtn").trigger("click");
        expect(wrapper.emitted("submit")).toEqual([[extraItems]]);
    });

});
