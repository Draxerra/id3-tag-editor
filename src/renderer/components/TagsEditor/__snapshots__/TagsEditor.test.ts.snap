// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags Editor Component snapshot has not changed 1`] = `
<form novalidate="novalidate" class="v-form TagsEditor">
  <div class="v-input TagsEditorTextField v-text-field v-input--is-label-active v-input--is-dirty">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot">
          <label for="name" class="v-label v-label--active" style="left: 0px; position: absolute;">Track Name</label>
          <input id="name" type="text">
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages">
          <div class="v-messages__wrapper"></div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-input TagsEditorTextField v-text-field v-input--is-label-active v-input--is-dirty">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot">
          <label for="artist" class="v-label v-label--active" style="left: 0px; position: absolute;">Artist</label>
          <input id="artist" type="text">
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages">
          <div class="v-messages__wrapper"></div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-input TagsEditorTextField v-text-field v-input--is-label-active v-input--is-dirty">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot">
          <label for="albumArtist" class="v-label v-label--active" style="left: 0px; position: absolute;">Album Artist</label>
          <input id="albumArtist" type="text">
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages">
          <div class="v-messages__wrapper"></div>
        </div>
      </div>
    </div>
  </div>
  <button type="button" class="TagsEditorSubmitBtn v-btn primary">
    <div class="v-btn__content">Save Changes</div>
  </button>
</form>
`;
